spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamBoundMessageConversionDelegate.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binder.kstream;
+
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+import org.springframework.cloud.stream.binder.kstream.config.KStreamBinderConfigurationProperties;
+import org.springframework.cloud.stream.converter.CompositeMessageConverterFactory;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.converter.MessageConverter;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.util.MimeType;
+import org.springframework.util.MimeTypeUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * @author Soby Chacko
+ */
+public class KStreamBoundMessageConversionDelegate {
+
+	private final CompositeMessageConverterFactory compositeMessageConverterFactory;
+
+	private final SendToDlqAndContinue sendToDlqAndContinue;
+
+	private final KStreamBinderConfigurationProperties kStreamBinderConfigurationProperties;
+
+	private final BoundedKStreamPropertyCache boundedKStreamPropertyCache;
+
+	public KStreamBoundMessageConversionDelegate(CompositeMessageConverterFactory compositeMessageConverterFactory,
+												SendToDlqAndContinue sendToDlqAndContinue,
+												KStreamBinderConfigurationProperties kStreamBinderConfigurationProperties,
+												BoundedKStreamPropertyCache boundedKStreamPropertyCache) {
+		this.compositeMessageConverterFactory = compositeMessageConverterFactory;
+		this.sendToDlqAndContinue = sendToDlqAndContinue;
+		this.kStreamBinderConfigurationProperties = kStreamBinderConfigurationProperties;
+		this.boundedKStreamPropertyCache = boundedKStreamPropertyCache;
+	}
+
+	public KStream serializeOnOutbound(KStream<?,?> outboundBindTarget) {
+		String contentType = this.boundedKStreamPropertyCache.getContentType(outboundBindTarget);
+		MessageConverter messageConverter = StringUtils.hasText(contentType) ? compositeMessageConverterFactory
+				.getMessageConverterForType(MimeType.valueOf(contentType))
```

Comment:
we can't do this in the constructor as this is set up per outbound and individual output bindings could use a different contentType.

Reply:
That's not what I meant. You could get converter for "all" types in the constructor and that will give you our CompositeMessageConverter with all converters we use and it will iterate through them and find the right one

Reply:
so this way if there are any "conversion issues" we are fixing them in one place *only*